src/main/java/fi/fmi/avi/converter/iwxxm/generic/XPathBuilder.java

```diff
@@ -0,0 +1,126 @@
+        final StringBuffer result = new StringBuffer();
+        while (matcher.find()) {
+            final String nsPattern = PREFIX_TO_NS_PATTERN.get(matcher.group(1));
+            final String localName = matcher.group(2);
+            final String replacement = nsPattern != null
+                    ? "/*[contains(namespace-uri(),'" + nsPattern + "') and local-name()='" + localName + "']"
+                    : "/*[local-name()='" + localName + "']";
+            matcher.appendReplacement(result, Matcher.quoteReplacement(replacement));
+        }
+        matcher.appendTail(result);
+
+        return result.toString();
+    }
+
+    /**
+     * Converts a human-readable XPath and wraps it for text extraction.
+     * The result selects the first matching node and extracts its normalized text content.
+     *
+     * @param readableXPath XPath with namespace prefixes
+     * @return version-agnostic XPath wrapped in normalize-space()
+     */
+    public static String text(final String readableXPath) {
+        return "normalize-space((" + toVersionAgnostic(readableXPath) + ")[1])";
+    }
+
+    /**
+     * Converts a human-readable XPath and wraps it for node selection.
+     * The result selects the first matching node.
+     *
+     * @param readableXPath XPath with namespace prefixes
+     * @return version-agnostic XPath selecting the first matching node
+     */
+    public static String node(final String readableXPath) {
+        return "(" + toVersionAgnostic(readableXPath) + ")[1]";
+    }
+
+    /**
+     * Converts a human-readable XPath and wraps it according to the field type.
+     *
+     * @param readableXPath XPath with namespace prefixes
+     * @param field         the field being queried, determines wrapping style
+     * @return version-agnostic XPath with appropriate wrapper
+     */
+    public static String wrap(final String readableXPath, final IWXXMField field) {
+        switch (field.getType()) {
+            case TEXT:
+                return text(readableXPath);
+            case NODE:
+                return node(readableXPath);
+            default:
+                throw new IllegalArgumentException("Unknown field type: " + field.getType());
+        }
+    }
+
+    /**
+     * Helper method for building field XPath maps in FieldXPathProvider implementations.
+     * Wraps each path according to the field's type and adds them to the map.
+     *
+     * @param map   the map to add the field expressions to
+     * @param field the field being mapped
+     * @param paths one or more XPath expressions for this field
+     */
+    public static void put(final Map<IWXXMField, List<String>> map, final IWXXMField field, final String... paths) {
+        final List<String> wrapped = new ArrayList<>(paths.length);
+        for (final String path : paths) {
+            wrapped.add(wrap(path, field));
+        }
+        map.put(field, wrapped);
```

Comment:
Probably safer to do

```suggestion
        map.put(field, Collections.unmodifiableList(wrapped));
```